HTTP unit tests (Reivew only, don't merge)

tgsong opened the pull request:
Description of changes:

- Turn on HTTP in unit tests config and add OTA HTTP unit tests
- Fix compiler warnings in OTA http
- Fix a bug in OTA http decode
- Add log level control for unit tests (so that compiler doesn't warn about variables only used in logs)
- Remove cmock from unit tests cmake file
- Add compiler warnings to CI checks
- Separate unit tests and coverage check (because in unit tests we want to turn on assert while in coverage we want it off)
- Clean up coverage data from cmake instead of in GitHub CI file

**Note: compiler warnings check is failing because of pending change on error enums.**

Commit:
Remove dependencies and unit tests source files from coverage

Commit:
Format coverage.cmake

Commit:
Add ota http to unit tests

Commit:
Fix compiler warnings in ota http

Commit:
Enable HTTP in config and fix some issues in ota http

- http decode file block should copy message to payload
- add const to some ota http interfaces

Commit:
Remove cmock from unit tests since we're not using it

Commit:
Add http unit tests and fix an issue in http decode

Commit:
Update CI

- add compiler warnings
- separate coverage and unit tests
- check branch coverage
- add log level to unit tests config

Commit:
Refactor ota unit tests

Commit:
Format files and update coverage CI

Commit:
Update lexicon.txt

divekarshubham commented on test/unit-test/ota_config.h:
```diff
@@ -31,6 +31,28 @@
+#define LOG_LEVEL_DEBUG                 3
+
+#ifndef LOG_LEVEL
+    #define LOG_LEVEL                   -1
```
Let the default value for log be LOG_LEVEL_ERROR ?

tgsong replied:
I find it bit noisy to turn on log when running unit tests since many are testing failing cases. I added this only for checking compiler warnings.

divekarshubham commented on test/unit-test/ota_utest.c:
```diff
@@ -864,7 +897,13 @@ void test_OTA_InitFileTransfer()
+void test_OTA_InitFileTransferHttp()
+{
+    pOtaJobDoc = JOB_DOC_HTTP;
+    test_OTA_InitFileTransferMqtt();
```
Can be refactored to:
```C
otaInitFileTransfer(){ ... }

test_OTA_InitFileTransferHttp(){ 
   pOtaJobDoc = JOB_DOC_HTTP;
   otaInitFileTransfer()
}

test_OTA_InitFileTransferMQTT(){ 
   pOtaJobDoc = JOB_DOC_A;
   otaInitFileTransfer()
}
```

tgsong replied:
Done

divekarshubham commented on test/unit-test/ota_utest.c:
```diff
@@ -877,6 +916,12 @@ void test_OTA_RequestFileBlock()
     TEST_ASSERT_EQUAL( OtaAgentStateWaitingForFileBlock, OTA_GetAgentState() );
 }
 
+void test_OTA_RequestFileBlockHttp()
```
Same as above.

Commit:
Refactor some tests

Commit:
Merge branch 'misra_warnings' into http_utest

Commit:
Fix unit tests because of API naming change

Commit:
Remove redundant test step in CI coverage

Commit:
Update coverage in CI

Commit:
MISRA warnings (#90)

* Adding OTA agent files

* Added tokens to lexicon.txt and spellings fix

* spell-check fix

* spell check and doxygen fix

* Added back OTA_JsonFileSignatureKey

* Support for coreJSON2.0

* Add fileType key support for multi file type support

* Fix filetype offset in job param table

* Update log strings in ota.c

* User buffer initial commit

* Remove unused code

* Alternative memory allocation support

* Implement request timer and self-test timers

* Cleanup on ota agent

* Update the ota config includes in ota.c and ota_private.h

* Fix issue with job parsing flag width

* Fix ota agent unit tests build errors

* Spell-check and formatting fix

* doxygen fix and added 2 error codes

* formatting and spell check fix

* Complexity reduction fix

* fixed commas for debug statements

* Remove void * from the json parsin table

* Add an error for failing to decode cbor

* Remove ota types header from library cmake paths

* Fix some compilation issues and warnings

* Fix PR feedfback in ota agent

* Fix CI checks

* Addressed PR feedback

* Fix pointer usage in job parsing (#76)

* Update lexicon and github workflow

* Sync fix from development

* Format files

* Sycn unit tests fix from development

See https://github.com/aws/ota-for-aws-iot-embedded-sdk/pull/70

* Format files

* Update comments in os posix unit tests

* Add missing files

* Turn on sanitizers when running unit tests

* Sycn OTA agent fix from development

* Fix complete callback initialization

* Fix compiler warnings

* Fix doxygen warnings

* Update lexicon.txt

* Convert buffer size to macros in unit tests

* Remove a magic number from ota mqtt

* Update c-sdk branch in github workflow

* Fix stdbool.h for msvc

* Enable more warnings in unit tests

* Fix some compiler warnings

- statements before declarations
- integer conversion
- missing returns
- typos in logs

* More compiler warnings fix

- unused parameters
- constness for job parsing
- printf specifiers

* Use const version of json search, format ota.h

* Fix misra warnings

* Silence misra warnings about snprintf in ota mqtt

* Suppress false positive misra warnings

* Sue pal interface and fix memory leak issues

* Add udpated ota private

* CI check fix

* Complexity reduction

* Bug fixes in #88 (#89)

* Fix include of stdbool.h

* Fix bugs in OTA agent

- decodeFileBlock should accept const network buffer
- decodeFileBlock_mqtt should not reuse network buffer
- remove unused methods
- signature buffer should not be dynamically allocated
- only update http url when it's not NULL
- decoded data payload should be freed after processDataBlock

* Format ota.c

* Fix compiler warnings

* Adress PR feedback

* Fix unit tests and bugs in OTA agent

* Lower the unit tests coverage to 80%

* Format files

* Reduce complexity increased due to PR feedback fixes

* Fix misra warnings

* Fix more misra warnings

* Fix compiler warnings

* Fix doxygen in ota mqtt

* Update lexicon.txt

* Reduce complexity

* Use different approach to test os posix timer

* Add detailed comments for suppressed MISRA warnings

* Update lexicon.txt

* Address PR feedback

* Correct spellings

* Capitalize posix in comments

Co-authored-by: Shubham Divekar <divekarshubham@gmail.com>
Co-authored-by: pvyawaha <pvyawaha@amazon.com>
Co-authored-by: Joshua Yan <52796499+yanjos-dev@users.noreply.github.com>

Commit:
Merge remote-tracking branch 'origin/main' into http_utest
